Add JavaScript linting

pedro-mendonca opened the pull request:
This is ~WIP~ finished.

Related to #1280 

- [x] Add linting script and GitHub action
- [x] Prepare ruleset with current errors related rules set to off
- [x] Re-enable rules
- [x] Fix JS coding standards

All rules implemented.

pedro-mendonca committed:
Add JavaScript linting with 'wp-scripts lint-js'

pedro-mendonca committed:
WPCS: Check "keyword-spacing"

pedro-mendonca committed:
WPCS: Fix "keyword-spacing"

pedro-mendonca committed:
WPCS: Check "space-before-blocks"

pedro-mendonca committed:
WPCS: Fix "space-before-blocks"

pedro-mendonca committed:
WPCS: Check "space-before-function-paren"

pedro-mendonca committed:
WPCS: Check "space-in-parens"

pedro-mendonca committed:
WPCS: Fix "space-in-parens"

pedro-mendonca committed:
WPCS: Check "no-mixed-spaces-and-tabs"

pedro-mendonca committed:
WPCS: Fix "no-mixed-spaces-and-tabs"

pedro-mendonca committed:
WPCS: Fix "no-multi-spaces"

pedro-mendonca committed:
WPCS: Fix "space-infix-ops"

pedro-mendonca committed:
WPCS: Fix "space-unary-ops"

pedro-mendonca committed:
WPCS: Fix "object-curly-spacing"

pedro-mendonca committed:
WPCS: Fix "computed-property-spacing"

pedro-mendonca committed:
WPCS: Fix "curly"

pedro-mendonca committed:
WPCS: Fix "quotes"

pedro-mendonca committed:
WPCS: Fix "padded-blocks"

pedro-mendonca committed:
WPCS: Fix "indent"

pedro-mendonca committed:
WPCS: Fix "one-var-declaration-per-line"

pedro-mendonca committed:
WPCS: Fix "semi"

pedro-mendonca committed:
WPCS: Fix "eqeqeq"

pedro-mendonca committed:
WPCS: Fix "brace-style"

pedro-mendonca committed:
WPCS: Fix "brace-style"

pedro-mendonca committed:
WPCS: Fix "comma-dangle"

pedro-mendonca commented:
@ocean90 can you please guide me through some of the next rules that I've muted?

Not sure if should continue to re-enable all of the muted checks.

I think we could at least turn them to "warn".

pedro-mendonca committed:
WPCS: Re-enable "no-mixed-operators"

ocean90 commented on .eslintrc.json:
```diff
@@ -0,0 +1,19 @@
+{
+  "extends": [ "plugin:@wordpress/eslint-plugin/recommended-with-formatting" ],
```
This seems to be the wrong config for this project since we are not using ESNext/React. For the start we should only use [es5](https://github.com/WordPress/gutenberg/blob/1e17b5a5b4a3117e843dde015c128361c5ae059e/packages/eslint-plugin/configs/es5.js). Can you change this to see which current disabled rules we can remove?

pedro-mendonca replied:
If I add only the *es5* the *i18n* is left out, should we add it also?

pedro-mendonca replied:
@ocean90  After applying this `es5` I see that the only rule I've fixed above that is not necessary on `es5` is the "computed-property-spacing". I can safely revert the fix on https://github.com/GlotPress/GlotPress-WP/pull/1281/commits/47743c8e0b9b764db613e7a03fa2b64e7d9a0e95

ocean90 replied:
`i18n` is only relevant if we use [`wp.i18n`](https://developer.wordpress.org/block-editor/reference-guides/packages/packages-i18n/). Since we don't we can safely ignore that for now.

> that is not necessary on `es5` is the "computed-property-spacing"

That's interesting since it's not really ESNext specific… Let's keep that and copy the custom rule for it:

```
'computed-property-spacing': [ 'error', 'always' ],
```

pedro-mendonca replied:
Done

pedro-mendonca committed:
Change JS cs ruleset to 'es5'

pedro-mendonca committed:
WPCS: Add muted rule "vars-on-top"

pedro-mendonca committed:
WPCS: Remove unnecessary rules

- no-var
- object-shorthand
- prefer-const
- @wordpress/no-unused-vars-before-return

pedro-mendonca committed:
Merge branch 'develop' into javascript-cs

pedro-mendonca committed:
WPCS: Keep "computed-property-spacing"

pedro-mendonca committed:
WPCS: Fix "no-else-return"

pedro-mendonca committed:
WPCS: Fix "indent"

pedro-mendonca committed:
WPCS: Re-enable "no-useless-return"

pedro-mendonca committed:
WPCS: Fix "wrap-iife"

pedro-mendonca committed:
WPCS: Fix "indent"

pedro-mendonca committed:
WPCS: Fix "indent"

pedro-mendonca committed:
WPCS: Fix "no-unused-expressions"

Can be tested clicking on Project Action to show and hide actions.

pedro-mendonca committed:
WPCS: Fix "no-unused-vars"

pedro-mendonca committed:
WPCS: Fix "vars-on-top"

On translations-page.js:39 added ignore to only set var inside current condition and avoid executing unnecessary code outside condition
On common.js added ignore to all occurrences

pedro-mendonca committed:
Sort rules alphabetically

pedro-mendonca commented:
@ocean90 from now I'll follow the example of the existent page top jscs comments and keep the matching exceptions.
After this, I think we can delete the jscs related files, right?

pedro-mendonca committed:
WPCS: Re-enable "no-alert" and added exception

pedro-mendonca committed:
WPCS: Re-enable "camelcase" and added exceptions

pedro-mendonca committed:
WPCS: Re-enable "no-undef" and added global and page definitions

pedro-mendonca committed:
WPCS: Set $gp globaly on .eslintrc.json

pedro-mendonca committed:
Remove 'jscs' page exceptions comments

pedro-mendonca committed:
WPCS: Define $gp per file again because is not needed on all files

pedro-mendonca committed:
WPCS: Re-enable "no-redeclare" and "no-shadow" as Warnings

ocean90 commented on assets/js/common.js:
```diff
@@ -1,81 +1,101 @@
+				focus: false,
+				group: 'default',
+			};
+			var options = $.extend( {}, defaults, options );
```
You can fix this by renaming the `options` argument to `userOptions`, `defaults` to `defaultOptions` and then use 

```suggestion
			var options = $.extend( {}, defaultOptions, userOptions );
```

pedro-mendonca replied:
"no-redeclare" fixed in https://github.com/GlotPress/GlotPress-WP/pull/1281/commits/834db775f2f19b83c38c4aae4b0438d096ea60bb

pedro-mendonca replied:
"no-shadow" fixed in https://github.com/GlotPress/GlotPress-WP/pull/1281/commits/125d7aab04747254c48777d4d9cbfe2a04cb8307

ocean90 reviewed (changes requested):
To fix `'data' is already declared in the upper scope` we could rename the argument in the `success()` callback to `response`.

pedro-mendonca committed:
WPCS: Fix "no-redeclare"

pedro-mendonca committed:
WPCS: Fix "no-shadow"

pedro-mendonca committed:
Build minified JS (UglifyJS)

Minified with grunt script

pedro-mendonca committed:
Merge branch 'develop' into javascript-cs

ocean90 commented:
@pedro-mendonca Since the title still includes WIP, is there anything left here or is good to merge?

pedro-mendonca commented:
@ocean90 it's done, I'm renaming it now.

pedro-mendonca commented:
@ocean90 are the below files still necessary?

- [.jscsrc](https://github.com/GlotPress/GlotPress-WP/blob/develop/.jscsrc)
- [.jshintignore](https://github.com/GlotPress/GlotPress-WP/blob/develop/.jshintignore)
- [.jshintrc ](https://github.com/GlotPress/GlotPress-WP/blob/develop/.jshintrc )
